### PR TITLE
feat: support question-answering model evaluation

### LINKED ISSUE #270: [Task] question-answering model support
## Overview

Extractive question-answering models take a context passage and a natural language question, then identify the answer span directly within the passage. These are reader/comprehension models (BERT, RoBERTa, DeBERTa, DistilBERT, ELECTRA variants) fine-tuned on SQuAD-style datasets.

The goal of this task is to verify that QA models exported through ModelKit (`wmk build`) produce accuracy that matches the HuggingFace PyTorch baseline — i.e., no accuracy regression from ONNX conversion or EP-specific quantization.

## Agent Scenarios

- **Document Q&A agent**: answer user questions grounded in a retrieved passage (contracts, manuals, reports) without hallucination — model reads the doc and extracts the answer span
- **Customer support agent**: resolve support queries by reading from a knowledge base article and returning the exact answer sentence
- **Research assistant agent**: extract specific facts (dates, names, figures) from long-form papers or reports on demand
- **On-device copilot agent**: run low-latency extractive QA locally on NPU for privacy-sensitive documents that cannot be sent to a cloud API

## ModelKit Integration

```
wmk config → wmk build (ONNX export) → wmk perf → wmk eval
```

Accuracy metric: Exact Match (EM) and F1 against SQuAD2.0, compared to HuggingFace PyTorch baseline. Regression threshold TBD.

## Acceptance Criteria

- [ ] ahotrod/electra_large_discriminator_squad2_512
- [ ] deepset/roberta-base-squad2
- [ ] google-bert/bert-large-uncased-whole-word-masking-finetuned-squad
- [ ] timpal0l/mdeberta-v3-base-squad2
- [ ] distilbert/distilbert-base-cased-distilled-squad
- [ ] deepset/bert-large-uncased-whole-word-masking-squad2
- [ ] monologg/koelectra-small-v2-distilled-korquad-384
- [ ] tau/splinter-base
- [ ] deepset/tinyroberta-squad2
- [ ] distilbert/distilbert-base-uncased-distilled-squad

### PR DESCRIPTION
**Pull request by zhenchaoni**
# Question Answering Evaluation — Design Spec

## Overview

The QA evaluator measures extractive question answering accuracy by comparing ONNX model predictions against a PyTorch baseline on SQuAD-format datasets. It wraps the HuggingFace `evaluate` library's `QuestionAnsweringEvaluator` and handles WinML-specific concerns (static ONNX shapes, SQuAD v2 auto-detection, optional input filtering).

**Evaluator class**: `WinMLQuestionAnsweringEvaluator` (extends `WinMLEvaluator`)
**Model class**: `WinMLModelForQuestionAnswering` (extends `WinMLPreTrainedModel`)

image

## HF Pipeline Output

The HuggingFace `QuestionAnsweringPipeline` performs extractive QA — it finds the answer span within the given context, rather than generating new text.

Given a question and a context passage, the pipeline tokenizes both, feeds them to the model which produces `start_logits` and `end_logits`, then extracts the highest-scoring text span from the context. For example, given question `"In what country is Normandy located?"` and a context containing `"...Normandy, a region in France."`, the model's logits peak at the token positions for "France", and the pipeline returns `{"answer": "France", "score": 0.73, "start": 94, "end": 100}`.

For SQuAD v2, the pipeline additionally computes `no_answer_probability` — when this exceeds a threshold, the predicted answer is empty (`""`).

### SQuAD v1

Every question has an answer span in the context. The predicted answer text is compared against the reference answer(s) using two metrics:

- **Exact Match (EM)**: Both texts are normalized (lowercase, strip articles/punctuation/whitespace), then compared. EM = 1 if the normalized prediction matches any reference answer exactly.
- **F1**: Both texts are split into word tokens, then bag-of-words precision and recall are computed. F1 = 2 * precision * recall / (precision + recall). When multiple reference answers exist, F1 is computed against each and the maximum is taken. This gives partial credit for partial matches — e.g., predicting "artificial intelligence" when the reference is "artificial intelligence and machine learning" yields F1 = 0.57 rather than 0.

### SQuAD v2

Extends v1 by adding unanswerable questions (~50% of the dataset), where the correct answer is empty. The `squad_v2` metric additionally handles no-answer predictions: if the model's `no_answer_probability` exceeds a threshold, the predicted answer is empty and compared accordingly. The metric reports both `HasAns` and `NoAns` breakdowns alongside the overall scores.

### F1 as the primary metric

F1 is used as the primary accuracy metric (`wmk_metric_key: "f1"`) because it provides partial credit and is more stable than EM. Like other accuracy metrics (classification accuracy, mAP), F1 is compared relatively between the quantized ONNX model and the PyTorch baseline — a relative delta under 5% is PASS, 5–10% is AT_RISK, over 10% is REGRESSION.

### Datasets

| Dataset | Format | Description |
|---------|--------|-------------|
| `rajpurkar/squad` (v1) | Answerable only | 10,570 validation samples |
| `rajpurkar/squad_v2` (v2) | Answerable + unanswerable | 11,873 validation samples |
| `KorQuAD/squad_kor_v1` | Korean, answerable only | 5,774 validation samples |

All share the same schema: `question`, `context`, `id`, `answers: {text: [str], answer_start: [int]}`.

### Evaluator: `WinMLQuestionAnsweringEvaluator`

- Delegates evaluation to HF `evaluator("question-answering")`
- **SQuAD v2 auto-detection**: `compute()` inspects the dataset via `is_squad_v2_format()` to check for unanswerable questions, then passes the matching `metric` (`"squad"` or `"squad_v2"`) and `squad_v2_format` flag to the HF evaluator
- Without this detection, HF defaults to the `squad` metric and crashes on v2 datasets because the pipeline produces `no_answer_probability` that the v1 metric doesn't expect
- `prepare_pipeline()` configures tokenizer padding to `max_length` matching the ONNX model's static sequence length (e.g., 512)

### Model: `WinMLModelForQuestionAnswering`

- Thin inference wrapper accepting `input_ids`, `attention_mask`, and `token_type_ids` from the HF pipeline
- Returns `QuestionAnsweringModelOutput` with `start_logits` and `end_logits`
- **Filters `token_type_ids`**: some architectures (e.g., RoBERTa, DeBERTa) do not include `token_type_ids` as an ONNX input because they use position-based segment encoding instead of explicit segment embeddings, but the HF tokenizer always produces and passes it
- Checks `self.io_config["input_names"]` and only forwards `token_type_ids` to the ONNX session if the model actually accepts it, preventing an ORT input mismatch error

### CODE REVIEW

**Review by DingmaomaoBJTU — CHANGES_REQUESTED** [review]
# PR #269 Code Review: Question Answering Model Evaluation

## Summary
This PR adds Question Answering (QA) task support to ModelKit's evaluation framework, including:
- `WinMLQuestionAnsweringEvaluator`: Evaluator for QA models on SQuAD datasets
- `WinMLModelForQuestionAnswering`: ONNX inference wrapper for QA models  
- Integration into the evaluation pipeline with 9 baseline models across v1/v2 SQuAD and Korean datasets

## Strengths ✓

1. **Clear design documentation** - The PR description thoroughly documents:
   - HuggingFace pipeline behavior (extractive QA, start/end logits)
   - SQuAD v1 vs v2 metric differences and auto-detection logic
   - Specific architectural considerations (token_type_ids filtering)

2. **Handles architecture variation** - Properly detects and filters `token_type_ids` for models that don't support it (RoBERTa, DeBERTa-v3)

3. **Comprehensive test coverage** - Baseline cache includes 9 diverse models across languages and dataset versions

4. **Proper schema definition** - `schema_info()` clearly documents expected column mappings (question, context, id, answers)

## Issues & Concerns ⚠️

### Critical

1. **Missing tests** ❌
   - No `tests/fixtures/create_test_models.py` changes shown  
   - No unit tests for `WinMLQuestionAnsweringEvaluator` or `WinMLModelForQuestionAnswering`
   - How is v2 auto-detection tested? What about edge cases?
   - CLAUDE.md rule: "Run `uv run pytest tests/` after every implementation or test revision"
   - **Action**: Add tests before merge

2. **Hardcoded column names** ❌
   - Lines 78-80 in evaluator use hardcoded defaults:
     ```python
     label_col = self.config.dataset.columns_mapping.get("label_column", "answers")
     ```
   - What if a dataset uses different naming? Should validate against schema
   - **Action**: Consider stricter validation or document assumptions

### Medium

3. **Import style inconsistency**
   - `question_answering_evaluator.py` line 12: Uses TYPE_CHECKING guard
   - `question_answering.py` line 18: Same guard
   - Good practice ✓ but ensure it's applied consistently across other evaluators

4. **Tokenizer padding logic** 
   - Lines 58-62: Extracts max_length from `io_config["input_shapes"][0][1]`
   - Assumes: (1) io_config exists, (2) input_shapes is populated, (3) sequence length is at index [0][1]
   - **Risk**: Silent failure if ONNX shape inference differs across models
   - **Suggestion**: Add validation with clear error messages

5. **Missing docstring for schema columns**
   - Lines 35-48: Good use of `SchemaColumn`, but `token_type_ids` isn't listed
   - Should document why it's auto-filtered and what models require it
   - **Action**: Add a note explaining optional inputs

### Minor

6. **Forward signature in question_answering.py**
   - Line 46-48: `**kwargs` is ignored for HF pipeline compat
   - Document which kwargs are ignored and why (better than silent drop)
   - Consider: Could any kwarg accidentally be critical?

7. **Dataset default in evaluate.py**
   - Line 94: Sets SQuAD v1 as default with 100 samples
   - Should document that v2 datasets auto-switch metrics
   - **Clarification needed**: Does the 100-sample default work for both v1 and v2?

8. **Error handling gaps**
   - What if `task_evaluator.is_squad_v2_format()` raises an error?
   - What if the pipeline returns malformed output?
   - **Suggestion**: Add try-catch with meaningful error context

## Code Quality Checklist

- ✅ Follows CLAUDE.md naming conventions (PascalCase for classes)
- ✅ Proper relative imports in `__init__.py` additions
- ✅ Copyright headers present
- ✅ Logging statements present
- ❌ **No tests added** (BLOCKER)
- ✅ Schema validation structure matches other evaluators
- ⚠️ Type hints present but incomplete (kwargs: Any is too loose)

## Questions for Author

1. **Test strategy**: Where are the unit tests? How do you validate SQuAD v2 auto-detection?
2. **Edge cases**: What happens if:
   - A model has NO token_type_ids input? 
   - The dataset has no "answers" column?
   - The ONNX shape is dynamic or missing?
3. **Backward compatibility**: Does this affect existing evaluation code paths?
4. **Performance**: What's the inference latency for the baseline models?

## Recommendation

**Request changes** - Merge pending:
1. Add unit tests for both new classes
2. Test SQuAD v1/v2 auto-detection with real datasets
3. Add error handling for missing io_config/input_shapes
4. Document token_type_ids filtering in schema comments
5. Run full pytest suite and confirm all pass

---

**Suggested commit message** (once fixed):
```
feat: add question answering evaluation with SQuAD v1/v2 support

- WinMLQuestionAnsweringEvaluator with automatic v2 format detection
- WinMLModelForQuestionAnswering wrapper filtering optional inputs
- 9 baseline models across SQuAD v1, v2, and Korean datasets
- F1 as primary metric with partial-credit scoring
```

File: src/winml/modelkit/models/winml/question_answering.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    Pipeline execution is done by WinMLAutoModel factory.
+    """
+
+    def forward(
+        self,
+        input_ids: torch.Tensor | np.ndarray,
+        attention_mask: torch.Tensor | np.ndarray | None = None,
+        token_type_ids: torch.Tensor | np.ndarray | None = None,
+        **kwargs: Any,
+    ) -> QuestionAnsweringModelOutput:
```
[the remark this answers is not in the document]
Comment (zhenchaoni) [reply src/winml/modelkit/models/winml/question_answering.py:49]: fixed

**Review by DingmaomaoBJTU — APPROVED** [review]
